### DISCUSSION TITLE
Reuse fields via inheritance

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#7511] [Usage Questions] [answered]
I wrote something like:

```
class Rhyme(Base):
    __tablename__ = "rhymes"

    id = sa.Column(sa.Integer, primary_key=True)
    term = sa.Column(sa.String(100), nullable=False, index=True)
    ...

class RhymeUS(Rhyme):
    __tablename__ = "rhymes_us"

    id = sa.Column(sa.Integer, sa.ForeignKey(Rhyme.id), primary_key=True)
```

The intention was to create a new table with the same fields (and indexes/constraints), but with the ID being a foreign key to the original table. However `RhymeUS.__table__` appears to only have the single field. Should I be doing something different?

Ultimately, these will be queried in 3 ways:
  Results of `Rhyme` only.
  `RhymeUS` together with `Rhyme` (so just a join query).
  And `RhymeUS` if it exists for each id, falling back to `Rhyme` if not (so `Rhyme` is basically the default which will be overridden by `RhymeUS`).

**Comment by zzzeek** [comment]
there's no automatic way to make a copy of a class mapping / table with the same columns, you would need to make a function that generates new declarative classes with the columns you want to have up front.

like


```py

def make_class():
   return type("MyClass", (Base, ), {"__tablename__": "mytable", "a": Column(Integer, primary_key=True), "b": Column(...)}))

```

**Reply by CaselIT** [reply]
can't a abstract mixing be used as well?

**Reply by Dreamsorcerer** [reply]
Looks tricky to work with mypy though. I might just leave it with copy/paste for now then.

**Reply by CaselIT** [reply]
I meant this section of the docs: https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html

I think you would probably need to define a `RhymeBase` that is extended by both classes

**Reply by Dreamsorcerer** [reply]
Yep, I was responding to the original answer.

It seems like the columns work as expected using `declarative_mixin`, but indexes/constraints give me `Can't add unnamed column to column collection`. e.g. with `__table_args__ = (sa.Index("idx_rhyme_1", ph1, syl),)`

**Reply by CaselIT** [reply]
if would need to use a method for `__table_args__ ` decorated with `declared_attr`, similar to `__tablename__` in the docs

**Reply by Dreamsorcerer** [reply]
Thanks, that looks promising now.

**Comment by Dreamsorcerer** [comment, marked as answer]
Resulting solution looks like:
```
@sa.orm.declarative_mixin
class _Rhyme:
    term = sa.Column(sa.String(100), nullable=False, index=True)
    ...

    @sa.orm.declared_attr
    def __table_args__(cls):
        return (sa.Index(cls.__tablename__ + "_my_index", cls.term, ...), ...)

class Rhyme(Base, _Rhyme):
    __tablename__ = "rhymes"

    id = sa.Column(sa.Integer, primary_key=True)

class RhymeUS(Base, _Rhyme):
    __tablename__ = "rhymes_us"

    id = sa.Column(sa.Integer, sa.ForeignKey(Rhyme.id), primary_key=True)
```
